ca_gazebo/src/create_virtual_wall_plugin.cpp

```diff
@@ -0,0 +1,147 @@
+
+static const char create2_model_name_prefix[] = "irobot_create2.";
+static const size_t create2_model_name_prefix_length = sizeof(create2_model_name_prefix) - 1;
+
+using gazebo::VirtualWallSensorPlugin;
+
+GZ_REGISTER_SENSOR_PLUGIN(VirtualWallSensorPlugin)
+
+void VirtualWallSensorPlugin::Load(gazebo::sensors::SensorPtr sensor, sdf::ElementPtr sdf)
+{
+  this->sensor_ = std::dynamic_pointer_cast<sensors::RaySensor>(sensor);
+  this->world_ = physics::get_world();
+  this->sensor_->SetActive(true);
+  ROS_DEBUG_NAMED("virtual_wall_plugin", "Loading");
+
+  if (!ros::isInitialized()) {
+      int argc = 0;
+      char** argv = NULL;
+      ros::init(argc,argv,"bumper_node",ros::init_options::AnonymousName);
+  }
+  this->rosnode_ = std::make_unique<ros::NodeHandle>();
+  this->updateNewEntity_ = event::Events::ConnectAddEntity(
+    std::bind(&VirtualWallSensorPlugin::OnAddEntity, this));
+  this->OnAddEntity();
+  // Listen to the update event. This event is broadcast every
+  // simulation iteration.
+  this->updateConnection_ = event::Events::ConnectWorldUpdateBegin(
+      std::bind(&VirtualWallSensorPlugin::OnUpdate, this));
+  ROS_DEBUG_NAMED("virtual_wall_plugin", "Loaded");
+}
+
+void VirtualWallSensorPlugin::OnAddEntity()
+{
+  const auto & world_models = this->world_->Models();
+  // Filter models which name starts with `irobot_create2.` and are continued with a number.
+  create_2_models_.resize(0);
+  pubs_.resize(0);
+  std::copy_if(
+    world_models.begin(),
+    world_models.end(),
+    std::back_inserter(this->create_2_models_),
+    [this] (const physics::ModelPtr & model)
+    {
+      const std::string & name = model->GetName();
+      ROS_DEBUG_NAMED("virtual_wall_plugin", "checking model %s", name.c_str());
+      if (name.size() < create2_model_name_prefix_length + 1) {
+        ROS_DEBUG_NAMED("virtual_wall_plugin", "model name is short");
+        return false;
+      }
+      if (std::strncmp(
+        name.c_str(),
+        create2_model_name_prefix,
+        create2_model_name_prefix_length) != 0)
+      {
+        ROS_DEBUG_NAMED("virtual_wall_plugin", "model name doesn't start with prefix");
+        return false;
+      }
+      ROS_DEBUG_NAMED("virtual_wall_plugin", "%s", name.substr(create2_model_name_prefix_length).c_str());
+      std::istringstream iss(name.substr(create2_model_name_prefix_length));
+      size_t x;
+      iss >> x;
+      if (iss) {
+        std::ostringstream oss;
+        oss << "create" << x << "/virtual_wall";
+        ROS_DEBUG_NAMED("virtual_wall_plugin", "model name ok");
+        this->pubs_.emplace_back(
+          this->rosnode_->advertise<std_msgs::Bool>(
+            oss.str().c_str(), 1));
+        return true;
+      }
+      ROS_DEBUG_NAMED("virtual_wall_plugin", "model name doesn't end with a number");
+      return false;
+    });
+  ROS_DEBUG_NAMED("virtual_wall_plugin", "number of create2: %zu", create_2_models_.size());
+}
+
+void VirtualWallSensorPlugin::OnUpdate()
```

Comment:
Use a loop to publish messages at a certain rate (specified in an SDF parameter). More info [here](http://answers.gazebosim.org/question/2058/how-to-control-publish-rate-in-ros-enabled-plugins/?answer=2059#post-id-2059).